L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+  // Returns true if verification occurs synchronously, in which case the
+  // verification result is populated in the sync_status output parameter.
+  // Returns false if verification occurs asynchronously, in which case the
+  // callback must be called when verification is complete and the verification
+  // result must be populated in the sync_status output parameter.
+  //
+  // request: the verification information associated with this request
+  // callback: This will only take effect if the verifier is asynchronous.
+  //           The function that gRPC will invoke when the verifier has already
+  //           completed its asynchronous check. Callers can use this function
+  //           to perform any additional checks. The input parameter of the
+  //           std::function indicates the status of the verifier check.
+  // sync_status: This will only be useful if the verifier is synchronous.
+  //              The status of the verifier as it has already done it's
+  //              synchronous check.
+  // return: return true if executed synchronously, otherwise return false
+  virtual bool Verify(TlsCustomVerificationCheckRequest* request,
+              std::function<void(Status)> callback,
+              Status* sync_status) = 0;
+
+  // Cancels a verification request previously started via Verify().
+  // Used when the connection attempt times out or is cancelled while an async
+  // verification request is pending.
+  //
+  // request: the verification information associated with this request
+  virtual void Cancel(TlsCustomVerificationCheckRequest* request) = 0;
+};
+
+// A CertificateVerifier that doesn't perform any additional checks other than
+// certificate verification, if specified.
+// Note: using this solely without any other authentication mechanisms on the
+// peer identity will leave your applications to the MITM(Man-In-The-Middle)
+// attacks. Users should avoid doing so in production environments.
+class NoOpCertificateVerifier : public CertificateVerifierInterface {
+ public:
+  NoOpCertificateVerifier();
+};
+
+// A CertificateVerifier that will perform hostname verification, to see if the
+// target name set from the client side matches the identity information
+// specified on the server's certificate.
+// This will be used as the default verifier if no verification is otherwise
+// set.
+class HostNameCertificateVerifier : public CertificateVerifierInterface {
+ public:
+  HostNameCertificateVerifier();
+};
+```
+
+#### Custom Chain Building
+```c++
+class CustomChainBuilderInterface {
+public:
+ virtual ~CustomChainBuilderInterface() = default;
+
+ // Returns the verified DER-encoded certificate chain, ordered from leaf to
+ // root. Both the leaf and the root are included. Returns an error status if
+ // verification fails for any reason.
+ virtual absl::StatusOr<std::vector<absl::string_view>> BuildAndVerifyChain(const std::vector<absl::string_view>& peer_cert_chain_der) = 0;
```

Comment:
From review, need to have story for the async version ready to go.

Reply:
Ack, working on this, resolving this comment just to work on cleaning up outdated comments